File: core/java/src/main/java/org/aiddl/core/container/Container.java
package org.aiddl.core.container;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;

import org.aiddl.core.interfaces.Module;
import org.aiddl.core.interfaces.Observer;
import org.aiddl.core.representation.ReferenceTerm;
import org.aiddl.core.representation.Substitution;
import org.aiddl.core.representation.SymbolicTerm;
import org.aiddl.core.representation.Term;
import org.aiddl.core.representation.TupleTerm;
import org.aiddl.core.tools.Logger;

/** 
 * Implementation of {@link Container}. Main access point to read and write AIDDL entries.
 * Each entry belongs to a module (identified by a name term). Operations that do not 
 * state a module usually use the working module which, by default, is the module of the
 * first file loaded into the container. 
 * 
 * @author Uwe Köckemann
 */
public class Container {
	
	private Module workingModule = null;
	
	private Map<Term,Module> modules;
	private List<Module> moduleList;

	Map<Term,Map<Term,Term>> aliasLookup;
	Map<Term,Term> selfAliasLookup;
	
	boolean threadSafe = false;
	
	/**
	 * Create an empty container.
	 */
	public Container() {
		this(false);
	}
	
	/**
	 * Create an empty, optionally thread safe, container.
	 * @param threadSafe set to <code>true</code> to use thread safe data structures internally
	 */
	public Container( boolean threadSafe ) {
		if ( threadSafe ) {
			this.threadSafe = true;
			modules = new ConcurrentHashMap<>();
			moduleList = new CopyOnWriteArrayList<>();

			aliasLookup = new ConcurrentHashMap<>();// HashMap<>();
			selfAliasLookup = new ConcurrentHashMap<>();
		} else {
			modules = new HashMap<>();
			moduleList = new ArrayList<>();

			aliasLookup = new HashMap<>();// HashMap<>();
			selfAliasLookup = new HashMap<>();
		}
	}
	
	public void switchModuleToThreadSafe( SymbolicTerm modName ) {
		Module m = this.modules.get(modName);
		if ( m instanceof LocalModule ) {
			((LocalModule)m).switchToThreadSafe();
		}
	}
		

	public Entry getEntry( Term name ) {
		for ( Entry e : workingModule.getEntries() ) {
			Substitution s = e.getName().match(name);
			if ( s != null ) {
				return e.substitute(s);
			}
		}
		return null;
	}
	

	public Entry getEntry( Term modName, Term name ) {
		Module m = modules.get(modName);
		if ( m == null ) {
			System.err.println(modName);
			System.err.println(name);
			System.err.println(this.aliasLookup);
			System.err.println(this.modules);
			throw new IllegalArgumentException("Requesting entry from unknown module: " + modName);
		}
		for ( Entry e : m.getEntries() ) {
			Substitution s = e.getName().match(name);
			if ( s != null ) {
				return e.substitute(s);
			}
		}
		return null;
	}
	

	public Collection<Entry> getMatchingEntries( Term modulePattern, Term typePattern, Term namePattern ) {
		ArrayList<Entry> r = new ArrayList<Entry>();
		for ( Module m : this.moduleList ) {
			Substitution s_base = null;
			if ( modulePattern != null ) {
				s_base = modulePattern.match(m.getName());
			} else {
				s_base = new Substitution();
			}
			if ( s_base != null ) {
				for ( Entry e : m.getEntries() ) {
					Substitution s = s_base.copy();
					
					if ( typePattern != null ) {
						Substitution s_tmp = typePattern.match(e.getType());
						if ( s_tmp == null || !s.add(s_tmp)) {
							continue;
						}
					}
					if ( namePattern != null ) {
						Substitution s_tmp = namePattern.match(e.getName());
						if ( s_tmp == null || !s.add(s_tmp)) {
							continue;
						} else {
							r.add(e);
						}
					} else if ( s != null ) {
						r.add(e);
					}
				}
			}
		}
		return r;
	}
	

	public List<Term> getModuleNames() {
		List<Term> moduleList = new ArrayList<>();
		for ( Module m : this.moduleList ) {
			moduleList.add(m.getName());
		}
		return moduleList;
	}

	

	public void setEntry( Entry entry ) {
		if ( !entry.getName().resolve(this).equals(entry.getName()) ) {
			throw new IllegalArgumentException("Entry name " + entry.getName() + " has references. Make sure to resolve references before creating the entry." );
		}
		
		Entry prev = this.getEntry(entry.name);
		if ( prev != null ) {
			workingModule.removeEntry(prev);
		}
		workingModule.putEntry(entry);
	}
	

	public void setEntry( Term module, Entry entry ) {
		Module mSet = this.modules.get(module);
		if ( mSet == null ) {
			throw new IllegalArgumentException("Module " + module + " does not exist.");
		}
		if ( !entry.getName().resolve(this).equals(entry.getName()) ) {
			throw new IllegalArgumentException("Entry name " + entry.getName() + " has references. Make sure to resolve references before creating the entry." );
		}
		
		Entry prev = this.getEntry(entry.name);
		if ( prev != null ) {
			mSet.removeEntry(prev);
		}
		mSet.putEntry(entry);
	}
	

	public void copyEntry( SymbolicTerm module_a, Term a, SymbolicTerm module_b, Term b ) {
		Entry aEntry = this.getEntry(module_a, a);
		this.setEntry(module_b, new Entry(aEntry.getType(),aEntry.getName(), aEntry.getValue()));
	}


	public void deleteEntry(Entry entry) {
		workingModule.removeEntry(entry);
	}

	public void deleteEntry(Term module, Entry entry) {
		Module mSet = this.modules.get(module);

		if ( mSet == null ) {
			throw new IllegalArgumentException("Module " + module + " does not exist.");
		}
		mSet.removeEntry(entry);
	}


//	public void compileAndAdd(String s) {
//		Parser.parseString(s, this);
//	}
	

	public void export( Term moduleName, String filename ) {
		Module m = this.modules.get(moduleName);
		if ( m == null ) {
			throw new IllegalArgumentException("Requesting non-existing module: " +  moduleName);
		}
		FileWriter fstream;
		try {
			fstream = new FileWriter(filename);

			BufferedWriter out = new BufferedWriter(fstream);
			//out.write("(#mod self " + m.getName()+ ")" );
			out.write("\n");
			for ( Entry e : m.getEntries() ) {
				out.write("(\n");
				out.write(Logger.prettyPrint(e.getType(), 1));
				out.write("\n");
				out.write(Logger.prettyPrint(e.getName(), 1));
				out.write("\n");
				out.write(Logger.prettyPrint(e.getValue(), 1));
				out.write("\n)\n\n");
			}
			out.close();
			
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	

	public Term resolveReference( ReferenceTerm t ) {
		Term next = t; 
		Term current = t;
		do {
			if ( next instanceof ReferenceTerm ) {
				current = next;
				next = resolveReferenceOnce((ReferenceTerm)current);
				if ( current.equals(next) ) {
					return current;
				}
			} else {
				return next;
			}
		} while ( next != null );
		throw new IllegalArgumentException("Reference cannot be resolved: " + t + " refered module: " + t.asRef().getRefModule());
	}
	
	private Term resolveReferenceOnce( ReferenceTerm t ) {
		Term target = t.getRefTarget();
		if ( target instanceof ReferenceTerm ) {
			return target;
		}
		Term modName = t.getRefModule();
		Entry e =  this.getEntry(modName, target);
		if ( e == null ) {
			return null;
		}
		if ( e.getType().equals(Term.sym("#def")) ) {
			return t;
		}
		Term next = e.getValue();
		return next;
	}
	

	@Override
	public String toString() {
		StringBuilder sB = new StringBuilder();
		sB.append("Entries:\n");
		for ( Module m : this.moduleList ) {
			sB.append( "Module: " + m.getName() );
			if ( m == workingModule ) {
				sB.append(" <--- Working Module");
			}
			
			sB.append("\n");
			for ( Entry e : m.getEntries() ) {
				sB.append("\t");
				sB.append(e.toString());
				sB.append("\n");
			}
		}
		return sB.toString();
	}
	

	public Term getWorkingModule() {
		return this.workingModule.getName();
	}
	

	public void setWorkingModule( Term moduleName ) {
		Module m = this.modules.get(moduleName);
		if ( m == null ) {
			throw new IllegalAccessError("Trying use non-existing module " + moduleName + " as working module.");
		}
		this.workingModule = m;
	}


	public void addModule( Term name ) {
		LocalModule m = new LocalModule( name, this.threadSafe );
		if ( this.modules.putIfAbsent(name, m) == null ) {
			this.moduleList.add(m);
		}
		if ( workingModule == null ) {
			workingModule = m;
		}
	}
	

//	public void addProxyModule( SymbolicTerm name, SymbolicTerm nodeUri, SocketReceiver in, SocketSender out ) {
//		ProxyModule m = new ProxyModule(name, nodeUri, out, in);
//		if ( this.modules.putIfAbsent(name, m) == null ) {
//			this.moduleList.add(m);
//		}
//		if ( workingModule == null ) {
//			workingModule = m;
//		}
//		for ( Entry e : m.getEntries() ) {
//			if ( e.getType().equals(Term.sym("#mod") )) {
//				this.addModuleAlias(name, e.getName(), name);
//			} else if ( e.getType().equals(Term.sym("#req"))) {
//				Term refMod = e.getValue();
//				if ( refMod instanceof StringTerm ) {
//					refMod = Parser.getModuleFromFile(refMod.getStringValue());
//				}
//				this.addModuleAlias(name, e.getName(), refMod);
//			}
//		}
//	}
	
	public void removeModule( Term name ) {
		Module m = this.modules.get(name);
		this.modules.remove(name);
		this.moduleList.remove(m);
		if ( this.workingModule.equals(m) ) {
			this.workingModule = null;
		}
	}

	public void addModuleAlias(Term sourceModule, Term alias, Term targetModule) {
		if ( !aliasLookup.containsKey(sourceModule) ) {
			aliasLookup.put(sourceModule, new HashMap<>());
		}
		aliasLookup.get(sourceModule).put(alias, targetModule);
		if ( sourceModule.equals(targetModule) ) {
			this.selfAliasLookup.put(sourceModule, alias);
		}
	}


	public Term resolveModuleAlias(Term sourceModule, Term alias) {
		if ( this.aliasLookup.containsKey(sourceModule) ) {
			return this.aliasLookup.get(sourceModule).get(alias);
		} 
		return null;
	}
	

	public Term findSelfAlias(Term moduleName ) {
		return this.selfAliasLookup.get(moduleName);
	}


	public void registerObserver( Term module, Term entryName, Observer obs ) {
		Module m = this.modules.get(module);
		if ( m == null ) {
			throw new IllegalAccessError("Trying to add observer to non-existing module: " + module);
		}
		m.addObserver(entryName, obs);
	}
	
//
//	public void toggleFullFunctionNames( boolean flag ) {
//		for ( Module m : this.moduleList ) {
//			m.toggleFullFunctionNames();
//		}
//	}
	

	public void toggleNamespaces( boolean flag ) {
		Map<Term, Substitution> sub_map = new HashMap<>();
		for ( Module m : this.moduleList ) {
			Substitution ns_sub = m.getNamespaceSubstitution();
			if ( !flag ) 
				ns_sub = ns_sub.inverse();
			sub_map.put(m.getName(), ns_sub);
		}
		
		for ( Module m : this.moduleList ) {
			Collection<Entry> namespaces = this.getMatchingEntries(m.getName(), Term.sym("#namespace"), null);
			namespaces.addAll(this.getMatchingEntries(m.getName(), Term.sym("#nms"), null));
			Substitution s = new Substitution(true);
			for ( Entry ns : namespaces ) {
				Term ns_term = ns.getValue();
				if ( ns_term instanceof  TupleTerm ) {
					Term ns_name = ns_term.get(0);
					Term ns_table = this.resolveReference(ns_term.get(1).asRef());
					
//					System.out.println("Namespace: " + ns_name);
//					System.out.println("Table:\n" + Logger.prettyPrint(ns_table, 0));
					
					Substitution ns_sub = new Substitution();
					for ( Term entry : ns_table.asCollection() ) {
						ns_sub.add(entry.get(ns_name), entry.get(0));
					}
					
//					System.out.println(ns_sub);
					
					if ( !s.add(ns_sub) ) {
						System.err.println("Namespace term: " + ns_term);
						System.err.println("Substitution before: " + s);
						System.err.println("Substitution failed: " + ns_sub);
						throw new IllegalStateException("Name space conflict in module: " + m.getName() + "\nConsider using #req instead of #namespace and use references.");
					}
					
				} else {
					Term ns_mod_name = this.resolveModuleAlias(m.getName(), ns.getName());
					if ( !s.add(sub_map.get(ns_mod_name)) ) {
						System.err.println(sub_map);
						System.err.println(ns);
						System.err.println(s);
						System.err.println(ns_mod_name);
						System.err.println(sub_map.get(ns_mod_name));
						throw new IllegalStateException("Name space conflict in module: " + m.getName() + "\nConsider using #req instead of #namespace and use references.");
					}
				}
			}
			m.substitute(s);
		}
	}
}

File: core/scala/src/main/scala/org/aiddl/core/scala/container/Container.scala
package org.aiddl.core.scala.container

import org.aiddl.core.scala.function.DefaultFunctionUri.EVAL
import org.aiddl.core.scala.function.{Evaluator, Function, Verbose}
import org.aiddl.core.scala.representation.*
import org.aiddl.core.scala.util.logger.Logger

import java.io.PrintWriter
import java.util.logging.Level
import scala.collection.mutable
import scala.collection.mutable.{HashMap, Map}

/**
 * Main data structure of the AIDDL core. Contains modules (usually parsed from .aiddl files) and a registry of all
 * known functions.
 */
class Container extends Verbose {

    private val funReg: mutable.Map[Sym, Function] = new mutable.HashMap
    private val entReg: mutable.Map[Sym, mutable.Map[Term, Entry]] = new mutable.HashMap
    private val entList: mutable.Map[Sym, List[Entry]] = new mutable.HashMap
    private var modList: List[Sym] = List.empty
    private val obsReg: mutable.Map[Sym, mutable.Map[Term, List[Function]]] = new mutable.HashMap
    private val aliasReg: mutable.Map[(Sym, Sym), Sym] = new mutable.HashMap
    private val interfaceReg: mutable.Map[Sym, Term] = new mutable.HashMap
    private val specialTypes: Set[Term] = Set(Sym("#type"), Sym("#def"), Sym("#req"), Sym("#nms"), Sym("#namespace"), Sym("#interface"), Sym("#mod"))

    private var verboseFunctions: List[Verbose] = Nil

    Function.loadDefaultFunctions(this)

    override def logGetVerboseSubComponents: List[Verbose] = this.verboseFunctions

    /**
     * Check if the container has a function
     * @param uri unique name of the function
     * @return <code>true</code> if the function exists, <code>false</code> otherwise
     */
    def hasFunction(uri: Sym): Boolean = funReg.contains(uri)

    /**
     * Add a new function
     * @param uri the name of the function
     * @param f the function
     */
    def addFunction(uri: Sym, f: Function): Unit = {
        if (f.isInstanceOf[Verbose]) {
            this.verboseFunctions = f.asInstanceOf[Verbose] :: this.verboseFunctions
        }
        this.funReg.put(uri, f)
    }

    /**
     * Get a function if it exists.
     * @param uri the name of the function
     * @return option containing the function if it exists, <code>None</code> otherwise
     */
    def getFunction(uri: Sym): Option[Function] = funReg.get(uri)

    /**
     * Get a reference to a function if it exists.
     * @param uri the name of the function
     * @return option containing the function reference if it exists, <code>None</code> otherwise
     */
    def getFunctionRef(uri: Sym):Option[FunRef] = funReg.get(uri).flatMap( f => Some(FunRef(uri, f)) )

    /**
     * Get a function if it exists or a default function otherwise.
     * @param uri the name of the function
     * @param default default function
     * @return the function if it exists, <code>default</code> otherwise
     */
    def getFunctionOrDefault(uri: Sym, default: Function):Function = funReg.getOrElse(uri, default)

    /**
     * Get a reference to a function if it exists or a reference to a default function otherwise.
     * @param uri the name of the function
     * @param default default function reference
     * @return the function reference if it exists, <code>default</code> otherwise
     */
    def getFunctionRefOrDefault(uri: Sym, default: FunRef): FunRef = funReg.get(uri).flatMap(f => Some(FunRef(uri, f))) match {
        case Some(value) => value
        case None => default
    }

    /**
     * Get a function if it exists and throws an exception otherwise
     * @param uri the name of the function
     * @return the function
     */
    def getFunctionOrPanic(uri: Sym): Function = {
        funReg.get(uri) match
            case Some(f) => f
            case None =>
                val stringBuilder = new StringBuilder()
                funReg.foreach(uri => stringBuilder.append(s"\n\t$uri"))
                throw new IllegalArgumentException(s"Function not registered: $uri. The following URIs are known:${stringBuilder.toString()}")
    }

    /**
     * Get a reference to a function if it exists and throws an exception otherwise
     * @param uri the name of the function
     * @return the function
     */
    def getFunctionRefOrPanic(uri: Sym): FunRef =
        FunRef(uri, getFunctionOrPanic(uri))

    /**
     * Add an interface definition.
     * @param uri the name of the interface
     * @param it term that defines the interface
     */
    def addInterfaceDef( uri: Sym, it: Term ): Unit = {
        interfaceReg.update(uri, it)
    }

    /**
     * Get interface defined under uri
     * @param uri name of the interface
     * @return interface definition term
     */
    def interface(uri: Sym): Term =
        this.interfaceReg(uri)

    /**
     * Get the evaluator function
     * @return evaluator registered in this container
     */
    def eval: Evaluator =
        funReg(EVAL).asInstanceOf[Evaluator]

    /**
     * Add a new module if a module under the given name does not exist. If the module exists, there is no change.
     *
     * @param module the name of the module
     */
    def addModule(module: Sym, alias: Sym = Sym("self")): Unit = {
        if (!entReg.contains(module)) {
            entReg.put(module, new mutable.HashMap)
            entList.put(module, Nil)
            modList = module :: modList
            this.addModuleAlias(module, alias, module)
            this.setEntry(module, Entry(Sym("#mod"), alias, module))
        }
    }

    /**
     * Get a list of all registered module names
     * @return list of module names
     */
    def getModuleNames: List[Sym] = modList

    /**
     * Save a module to a file
     * @param module name of the module
     * @param name filename to write module to
     */
    def saveModule(module: Sym, name: String): Unit = {
        //val modLine = s"(#mod ${this.findSelfAlias(module)} $module)"
        val sb = new mutable.StringBuilder
        //sb.append(modLine)
        //sb.append("\n\n")

        this.entList(module)
          .reverse
          .foreach( e => {
            sb.append(Logger.prettyPrint(e.asTuple, 0))
            sb.append("\n\n")
        })

        new PrintWriter(name) { write(sb.toString()); close() }
    }

    private def callObservers(module: Sym, entry: Entry) = {
        val obs = obsReg.getOrElse(module, mutable.Map.empty).getOrElse(entry.name, List())
        obs.map(_ apply entry.value)
    }

    /**
     * Set an entry in a module. If the module does not exist, it will be created.
     * @param module name of the module
     * @param entry the entry to set
     * @return option value containing entry that was overwritten or <code>None</code> if an entry under this name did
     *         not exist before
     */
    def setEntry(module: Sym, entry: Entry): Option[Entry] = {
        addModule(module)
        callObservers(module, entry)
        (entReg.get(module): @unchecked) match { 
            case Some(m) =>
                m.get(entry.name) match {
                    case Some(e) => entList.put(module, entry :: entList(module).filter(x => x != e))
                    case _ => entList.put(module, entry :: entList(module))
                }
                //if ( m.contains(entry.n) ) {}
                m.put(entry.name, entry)
        }
    }

    /**
     * Delete an entry from a module
     * @param module the module
     * @param entry entry to delete
     * @return option value containing deleted entry if it existed, <code>None</code> otherwise
     */
    def deleteEntry(module: Sym, entry: Entry): Option[Entry] = {
        callObservers(module, entry)
        entReg.getOrElse(module, mutable.Map.empty).remove(entry.name)
    }

    /**
     * Get an entry by its name from a module.
     * @param module name of the module
     * @param name name of the entry
     * @return option value containing the entry, <code>None</code> otherwise
     */
    def getEntry(module: Sym, name: Term):Option[Entry] =
        entReg.get(module) match {
            case Some(m) => m.get(name)  match {
                case Some(e) => Some(e)
                case None => m.find( (k, e) => k unifiable name ).flatMap( t => Some(t(1) \ (t(0) unify name).get ) )
            }
            case None => None }

    /**
     * Get all entries with a matching name, module, and type
     * @param modulePattern pattern for the module name
     * @param typePattern pattern for the type
     * @param namePattern pattern for the name
     * @return list of all entries in the container that match all patterns
     */
    def getMatchingEntries(modulePattern: Term, typePattern:Term, namePattern: Term): List[Entry] = {
        entList.filter( { case (k, _) => modulePattern unifiable k } )
               .values
               .flatMap( m => m.filter( { case Entry(t, n, _) => (namePattern unifiable n) && (typePattern unifiable t) } ) )
               .toList.reverse
    }

    /**
     * Get the value of an entry after using the evaluator to process it and resolve all references
     * @param module name of the module
     * @param name name of the entry
     * @return option value containing the processed value, <code>None</code> otherwise
     */
    def getProcessedValue(module: Sym, name: Term): Option[Term] =
        this.getEntry(module, name).flatMap( e => Some(this.eval.evalAllRefs(this.resolve(e.value))) )
    /**
     * Get the value of an entry after using the evaluator to process it and resolve all references.
     * Throws an exception if the module does not contain an entry with the given name
     * @param module name of the module
     * @param name name of the entry
     * @return the processed value
     */
    def getProcessedValueOrPanic(module: Sym, name: Term): Term =
        this.getEntry(module, name) match {
            case Some(e) => this.eval.evalAllRefs(e.value)
            case None => throw new NoSuchElementException(s"Module $module does not have an entry named $name")
        }

    /**
     * Get all entries associated to a given module
     * @param m name of the module
     * @return list of entries in <code>m</code>
     */
    def getModuleEntries(m: Sym): List[Entry] = this.entList.getOrElse(m, Nil)

    /**
     * Resolve entry references in a term.
     * @param t the term
     * @return a term with all entry references replaced by the values of their entries
     */
    def resolve( t: Term ): Term = t match {
        case Tuple(args @ _*) => Tuple(args.map(x => this.resolve(x)): _*)
        case ListTerm(list) => ListTerm(list.map(x => this.resolve(x)))
        case SetTerm(set) => SetTerm(set.map(x => this.resolve(x)))
        case KeyVal(key, value) => KeyVal(this.resolve(key), this.resolve(value))
        case r @ EntRef(_, _, _) => resolve(resolveReference(r))
        case _ => t
    }

    /**
     * Resolve a single entry reference. Follows references chains until the result is no longer an entry reference term.
     * @param r entry reference
     * @return non entry reference term
     */
    def resolveReference( r: EntRef ): Term = {
        resolveReferenceOnce(r) match {
            case Some(next @ EntRef(_, _, _)) if next == r => next
            case None => throw new IllegalArgumentException(s"Cannot resolve reference: $r target=${r.mod} (class: ${r.getClass.getSimpleName})" )
            case Some(next @ EntRef(_, _, _)) => resolveReference(next)
            case Some(v) => v
        }
    }

    private def resolveReferenceOnce(r: EntRef): Option[Term] = {
        if r.name.isInstanceOf[EntRef] then Some(r.name)
        else
            this.getEntry(this.findModuleAlias(r.mod, r.alias), r.name) match
                case Some(Entry(Sym("#def"), _, _)) => Some(r)
                case Some(Entry(_, _, v)) => Some(v)
                case _ => None
    }

    /**
     * Add an alias for another module to a module
     * @param module module that uses the alias
     * @param alias the alias
     * @param target the target module
     */
    def addModuleAlias(module: Sym, alias: Sym, target: Sym): Unit = {
        this.aliasReg((module, alias)) = target
    }

    /**
     * Get the actual name for a module alias
     * @param source module that uses the alias
     * @param alias the alias
     * @return the global name referred to by the alias
     */
    def findModuleAlias(source: Sym, alias: Sym):Sym = {
        this.aliasReg.get((source, alias)) match {
            case Some(uri) => uri
            case None => throw new IllegalArgumentException(s"Alias $alias not defined in module $source.\nAliases are added with requirement entries such as: (#req $alias MODULE), where MODULE is the symbolic name of a module or a relative filename.")

        }
    }

    /**
     * Find the self-alias of a module. The self alias is established in the #mod entry of a module. If the #mod entry is
     * <code>(#mod self name)</code>, the module <code>name</code> refers to itself with the symbol <code>self</code>.
     * @param source the module
     * @return the module's self alias
     */
    def findSelfAlias(source: Sym):Sym =
        this.aliasReg.find((k, target) => k(0) == target).get(0)(1)

    /**
     * Check if entries have the correct types in all modules.
     * @param verbose flag to generate some debug output
     * @return <code>true</code> if all entry values satisfy their given types, <code>false</code> otherwise
     */
    def typeCheckAllModules(): Boolean = {
        modList.forall(m => {
            this.logger.info(s"Module: $m")
            typeCheckModule(m)
        })
    }

    /**
     * Type check all entries in a single module
     * @param uri name of the module
     * @param verbose turn on debug messages
     * @return <code>true</code> if all entry values satisfy their given types, <code>false</code> otherwise
     */
    def typeCheckModule(uri: Sym, verbose: Boolean = false): Boolean = {
        entList.get(uri) match {
            case Some(es) =>
                es.forall( e => {
                    val isConsistent = checkSingleType(uri, e.typeRef, e.value)
                    this.logger.info(s"$uri | ${e.typeRef} | ${e.name} | $isConsistent ")
                    isConsistent
                })
            case None =>
                val sb = new StringBuilder()
                sb.append(s"There are ${modList.length} registered modules:")
                modList.foreach(m => sb.append(s"\n\t$m"))
                throw new IllegalArgumentException(s"Module $uri does not exist.\n${sb.toString()}")
        }
    }

    private def checkSingleType(m: Sym, t: Term, v: Term): Boolean = {
        try {
            eval(t) match {
                case _: FunRef => t(this.eval(resolve(v))).asBool.v
                case s: Sym if this.specialTypes contains s => true
                case furi: Sym => this.getFunctionOrPanic(furi)(this.eval(resolve(v))).asBool.v
                case _ =>
                    val hint = if (!t.isInstanceOf[EntRef]) "" else s"\nHint: If this entry reference corresponds to a type, try using ^$t instead (adding the ^) to make it a function reference."
                    throw new IllegalArgumentException(s"Bad type specifier $t in module $m. Use symbolic type URI or function reference instead.$hint")
            }
        } catch {
            case ex: Throwable => {
                ex.printStackTrace()
                throw new IllegalArgumentException(s"Error when checking type ${t} in module $m with value $v.")
            }
        }
    }

    override def toString: String =
        this.modList
          .map(m => this.entList(m).reverse.mkString("", "\n", ""))
          .reverse
          .mkString("", "\n", "")
}

